File: src/twenty/december/no208/ImplementTrie.java
package twenty.december.no208;

public class ImplementTrie {
    public static void main(String[] args) {
        Trie trie = new Trie();
        trie.insert("apple");
        System.out.println(trie.search("apple"));   //true
        System.out.println(trie.search("app"));     //false
        System.out.println(trie.startsWith("app"));      //true
        trie.insert("app");
        System.out.println(trie.search("app"));     //true
    }
}

class Trie {

    private static final int MAX_SIZE = 26;
    private final Trie[] childrens = new Trie[MAX_SIZE];
    private boolean isEnd = false;

    /**
     * Initialize your data structure here.
     */
    public Trie() {
    }

    /**
     * Inserts a word into the trie.
     */
    public void insert(String word) {
        Trie node = this;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (!node.containsKey(c)) {
                node.put(c, new Trie());
            }
            node = node.get(c);
        }
        node.setEnd();
    }

    /**
     * Returns if the word is in the trie.
     */
    public boolean search(String word) {
        Trie trie = searchPrefix(word);
        return trie != null && trie.isEnd();
    }

    /**
     * Returns if there is any word in the trie that starts with the given prefix.
     */
    public boolean startsWith(String prefix) {
        return searchPrefix(prefix) != null;
    }

    private Trie searchPrefix(String prefex) {
        Trie node = this;
        for (int i = 0; i < prefex.length(); i++) {
            char c = prefex.charAt(i);
            if (!node.containsKey(c)) {
                return null;
            }
            node = node.get(c);
        }
        return node;
    }

    private void put(char c, Trie node) {
        childrens[c - 'a'] = node;
    }

    private Trie get(char c) {
        return childrens[c - 'a'];
    }

    private boolean containsKey(char c) {
        return childrens[c - 'a'] != null;
    }

    public boolean isEnd() {
        return isEnd;
    }

    public void setEnd() {
        isEnd = true;
    }
}

/*
* 实现一个 Trie (前缀树)，包含 insert, search, 和 startsWith 这三个操作。

示例:

Trie trie = new Trie();

trie.insert("apple");
trie.search("apple");   // 返回 true
trie.search("app");     // 返回 false
trie.startsWith("app"); // 返回 true
trie.insert("app");
trie.search("app");     // 返回 true
说明:

你可以假设所有的输入都是由小写字母 a-z 构成的。
保证所有输入均为非空字符串。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/implement-trie-prefix-tree
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month3/no208/ImplementTrie.go
package main

func main() {
	trie := Constructor()
	//trie.Insert("apple")
	//println(trie.Search("apple"))   //true
	//println(trie.Search("app"))     //false
	//println(trie.StartsWith("app")) //true
	//trie.Insert("app")
	println(trie.Search("app")) //true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */

type Trie struct {
	childs [26]*Trie // 只有小写字母，因此用长度为 26 的数组存储子节点
	word   string    // 用来存储单词，用于 Search 方法标记当前节点是否为一个单词（一个叶子节点）
}

func Constructor() Trie {
	return Trie{}
}

func (this *Trie) Insert(word string) {
	p := this
	for _, c := range word {
		index := c - 'a'
		if p.childs[index] == nil {
			p.childs[index] = &Trie{}
		}
		p = p.childs[index]
	}
	p.word = word
}

func (this *Trie) getByPrefix(prefix string) (res *Trie) {
	p := this
	for _, c := range prefix {
		index := c - 'a'
		if p.childs[index] == nil {
			return nil
		}
		p = p.childs[index]
	}
	return p

}

func (this *Trie) Search(word string) bool {
	p := this.getByPrefix(word)
	return p != nil && p.word == word
}

func (this *Trie) StartsWith(prefix string) bool {
	return this.getByPrefix(prefix) != nil
}

/*
Trie（发音类似 "try"）或者说 前缀树 是一种树形数据结构，用于高效地存储和检索字符串数据集中的键。这一数据结构有相当多的应用情景，例如自动补全和拼写检查。

请你实现 Trie 类：
Trie() 初始化前缀树对象。
void insert(String word) 向前缀树中插入字符串 word 。
boolean search(String word) 如果字符串 word 在前缀树中，返回 true（即，在检索之前已经插入）；否则，返回 false 。
boolean startsWith(String prefix) 如果之前已经插入的字符串 word 的前缀之一为 prefix ，返回 true ；否则，返回 false 。


示例：
输入
["Trie", "insert", "search", "search", "startsWith", "insert", "search"]
[[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]
输出
[null, null, true, false, true, null, true]

解释
Trie trie = new Trie();
trie.insert("apple");
trie.search("apple");   // 返回 True
trie.search("app");     // 返回 False
trie.startsWith("app"); // 返回 True
trie.insert("app");
trie.search("app");     // 返回 True

提示：
1 <= word.length, prefix.length <= 2000
word 和 prefix 仅由小写英文字母组成
insert、search 和 startsWith 调用次数 总计 不超过 3 * 104 次
*/
